### PR TITLE
Update resource prefixes for bound CRDs based on APIExport identity

### PR DESCRIPTION
**Pull request by ncdc**
Needed for https://github.com/kcp-dev/kcp/pull/877

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	if _, bound := t.crd.Annotations["apis.kcp.dev/bound-crd"]; !bound {
+		return ret, nil
+	}
+
+	apiIdentity := t.crd.Annotations["apis.kcp.dev/identity"]
+	if apiIdentity == "" {
+		return generic.RESTOptions{}, fmt.Errorf("missing 'apis.kcp.dev/identity' annotation")
+	}
+
+	ret.ResourcePrefix += "-" + apiIdentity
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go]: nit: resource prefixes of CRDs cannot have a dash? I.e. there is no risk of an attack hijacking data this way by creating a normal CRD with prefix+apiIdentity as the prefix?

**Comment by sttts** [comment]
Lgtm modulo the answer to https://github.com/kcp-dev/kubernetes/pull/57#discussion_r853083998

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go (hunk 1 above)
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go]: Good catch. Yes they can. I tested with a resource name of `dashed-endpoints` in the APIResourceSchema is this is what is in etcd when you create one (via an APIBinding):

```
/registry/andy.io/dashed-endpoints-d9d18877572da255c263f0dfcb0e39f5a7cbf7177c68f8dec5561ec81611d3e5/root:default:andy/default/test1
```

Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go]: Looks like the plural has to be DNS1035 (lowercase alphanumeric or `-`, start with alphabetic, end with alphanumeric)

Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go]: So maybe we use `:`?

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go]: : is fine, if that is valid (in URLs it is, in etcd too).

Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go]: : works in etcd. I don't believe that it's relevant to URLs (the `:<identity>` is not user-facing or in any URLs)

**Comment by ncdc** [comment]
Updated to use `:`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/registry/generic"
+)
+
+type apiBindingAwareCRDRESTOptionsGetter struct {
+	delegate generic.RESTOptionsGetter
+	crd      *apiextensionsv1.CustomResourceDefinition
+}
+
+func (t apiBindingAwareCRDRESTOptionsGetter) GetRESTOptions(resource schema.GroupResource) (generic.RESTOptions, error) {
```
Comment (ncdc) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go:15]: Note, any error we return from this function will **panic** the apiserver (https://github.com/kcp-dev/kubernetes/blob/feature-logical-clusters-1.23/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go#L108)

**Comment by sttts** [comment]
Lgtm

**Pull request merged**